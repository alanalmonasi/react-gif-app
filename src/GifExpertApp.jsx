import { useState } from 'react';
import { AddCategory } from './components/AddCategory';
import { GifGrid } from './components/GifGrid'

export const GifExpertApp = () => {

   const [categories, setCategories] = useState([]);

   const addCategory = (newCategory) => {

      if (categories.includes(newCategory)) return;

      setCategories([newCategory, ...categories]);
   }

   return (
      <>
         <h1>My GifExpertApp</h1>

         <AddCategory
            onNewValue={(inputValue) => addCategory(inputValue)} 
         />

         {
            categories.map(category => (
               <GifGrid 
                  key={category} 
                  category={category}
               />
            ))
         }
      </>
   )
}
