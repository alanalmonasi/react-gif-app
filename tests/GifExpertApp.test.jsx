import { render, screen, fireEvent } from '@testing-library/react';
import { GifExpertApp } from '../src/GifExpertApp'

describe('testing GifExpertApp', () => {

   const inputValue = 'GOKU';

   test('h1 should exist', () => {

      render(<GifExpertApp />)

      const h1 = screen.getByRole('heading', { level: 1 });

      expect(h1.innerHTML).toBe('My GifExpertApp');

   });

   test('should match the snapshot', () => {

      const { container } = render(<GifExpertApp />);

      expect(container).toMatchSnapshot();
   });

   test('should show input.value on DOM with value we are passing', () => {
      
      render(<GifExpertApp />)

      const input = screen.getByRole('textbox');
      fireEvent.input(input, { target: { value: inputValue } });

      expect(input.value).toBe('GOKU')

      //screen.debug();
   });

   test('should addCategory and match the input', () => {
      render(<GifExpertApp />);

      const input = screen.getByRole('textbox');
      const form = screen.getByRole('form');

      fireEvent.input(input, { target: { value: inputValue } });
      fireEvent.submit(form);
      
      expect(screen.getByRole('heading', { level: 3})).toBeTruthy();
      expect(screen.getByRole('heading', { level: 3}).innerHTML).toBe(inputValue);
   });

   test('should add category if it doesnt exist', () => {
      render(<GifExpertApp />);

      const input = screen.getByRole('textbox');
      const form = screen.getByRole('form');

      fireEvent.input(input, { target: { value: inputValue } });
      fireEvent.submit(form);

      fireEvent.input(input, { target: { value: 'CHEEMS' } });
      fireEvent.submit(form);

      const item = screen.getAllByRole('heading', { level: 3 }).length;

      expect(item).toBeGreaterThan(1);

      //screen.debug();
   });

   test('should not add category if it already exists', () => {
      render(<GifExpertApp />);

      const input = screen.getByRole('textbox');
      const form = screen.getByRole('form');

      fireEvent.input(input, { target: { value: inputValue } });
      fireEvent.submit(form);
      
      fireEvent.input(input, { target: { value: 'MIKASA' } });
      fireEvent.submit(form);

      fireEvent.input(input, { target: { value: inputValue } });
      fireEvent.submit(form);

      //screen.debug();

      const items = screen.getAllByRole('heading', { level: 3 });

      expect(items.length).toBe(2);
   })
});