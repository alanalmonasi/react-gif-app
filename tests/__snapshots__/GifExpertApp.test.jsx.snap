// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing GifExpertApp should match the snapshot 1`] = `
<div>
  <h1>
    My GifExpertApp
  </h1>
  <form
    aria-label="form"
  >
    <input
      placeholder="Buscar gif"
      type="text"
      value=""
    />
  </form>
</div>
`;
